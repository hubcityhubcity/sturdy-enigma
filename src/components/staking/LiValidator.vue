<template>
  <tr
    class="li-validator"
    :data-name="validator.name"
    @click="
      $router.push({
        name: 'validator',
        params: { validator: validator.operatorAddress }
      })
    "
  >
    <td>{{ index + 1 }}</td>
    <td class="hide-xs">
      <div class="status-container">
        <span
          :class="validator.status | toLower"
          class="validator-status"
          :title="validator.statusDetailed"
        >
          {{ validator.status }}
        </span>
      </div>
    </td>
    <td class="data-table__row__info">
      <Avatar
        v-if="!validator || !validator.picture || validator.picture === 'null'"
        class="li-validator-image"
        alt="generic validator logo - generated avatar from address"
        :address="validator.operatorAddress"
      />
      <img
        v-else-if="validator && validator.picture"
        :src="validator.picture"
        class="li-validator-image"
        :alt="`validator logo for ` + validator.name"
      />
      <div class="validator-info">
        <h3 class="li-validator-name">
          {{ validator.name }}
        </h3>
        <div v-if="delegation.amount > 0">
          <h4>
            {{ delegation.amount | shortDecimals }}
          </h4>
          <h5 v-if="rewards.amount > 0.001">
            +{{ rewards.amount | shortDecimals }}
          </h5>
        </div>
      </div>
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'expectedReturns' }">
      {{
        validator.expectedReturns ? percent(validator.expectedReturns) : `--`
      }}
    </td>
    <td :class="{ 'hide-xs': showOnMobile !== 'voting-power' }">
      {{ validator.votingPower | percent }}
    </td>
  </tr>
</template>

<script>
import { percent, shortDecimals, atoms } from "scripts/num"
import Avatar from "common/Avatar"
export default {
  name: `li-validator`,
  components: {
    Avatar
  },
  filters: {
    atoms,
    shortDecimals,
    percent,
    toLower: text => text.toLowerCase()
  },
  props: {
    validator: {
      type: Object,
      required: true
    },
    delegation: {
      type: Object,
      default: () => ({})
    },
    rewards: {
      type: Object,
      default: () => ({})
    },
    index: {
      type: Number,
      required: true
    },
    showOnMobile: {
      type: String,
      /* istanbul ignore next */
      default: () => "returns"
    }
  },
  methods: {
    percent
  }
}
</script>
<style scoped>
.li-validator {
  padding: 0.5rem 1rem;
  margin-bottom: 0.25rem;
  border-bottom: 1px solid var(--bc-dim);
  border-radius: 0.25rem;
}
.li-validator:last-child {
  border-bottom: none;
}
.validator-info {
  display: flex;
  flex-direction: column;
  padding-left: 1rem;
  text-overflow: ellipsis;
}
.li-validator h4,
.li-validator h5 {
  font-size: var(--sm);
  display: inline-block;
}
.li-validator h5 {
  padding-left: 0.5rem;
  color: var(--success);
}
.li-validator:hover {
  cursor: pointer;
  background: var(--hover-bg);
  color: var(--bright);
}
.li-validator-name {
  font-size: 1rem;
  line-height: 18px;
  font-weight: 500;
  color: var(--bright);
  display: inline-block;
  max-width: 20rem;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}
.li-validator-image {
  border-radius: 0.25rem;
  height: 2.5rem;
  width: 2.5rem;
  min-height: 2.5rem;
  min-width: 2.5rem;
  border: 1px solid var(--bc-dim);
}
.validator-status {
  text-transform: uppercase;
  font-size: 10px;
  font-weight: 600;
  border: 2px solid;
  padding: 2px 4px;
  border-radius: 0.25rem;
}
.validator-status.inactive {
  color: var(--warning);
  border-color: var(--warning);
}
.validator-status.active {
  color: var(--success);
  border-color: var(--success);
}
@media screen and (max-width: 768px) {
  .li-validator-name {
    max-width: 11rem;
  }
}
</style>
