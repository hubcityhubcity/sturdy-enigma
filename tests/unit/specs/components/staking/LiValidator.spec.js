import { shallowMount, createLocalVue } from "@vue/test-utils"
import VueApollo from "vue-apollo"
import LiValidator from "src/components/staking/LiValidator"

const localVue = createLocalVue()
localVue.directive(`tooltip`, () => {})
localVue.use(VueApollo)

describe(`LiValidator`, () => {
  let wrapper, $store

  const validator = {
    //From GraphQL
    avatarUrl: "https://s3.amazonaws.com/img/path.jpg",
    consensus_pubkey: "cosmosvalpub1234",
    customized: false,
    delegator_shares: "697935712090.000000000000000000",
    details: "Mr Mounty",
    identity: "4BE49EABAA41B8BF",
    jailed: false,
    keybaseId: "4BE49EABAA41B8BF",
    lastUpdated: "2019-08-15T16:03:34.988007+00:00",
    max_change_rate: "0.010000000000000000",
    max_rate: "0.500000000000000000",
    min_self_delegation: "1",
    moniker: "mr_mounty",
    operator_address: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au",
    profileUrl: "https://keybase.io/melea",
    rate: "0.080000000000000000",
    status: 2,
    tokens: 14,
    unbonding_height: 556130,
    unbonding_time: "2019-06-26T23:50:09.38840326Z",
    update_time: "2019-04-01T17:07:19.892771401Z",
    uptime_percentage: "0.9962",
    userName: "mr_mounty",
    website: "www.monty.ca",
    voting_power: 0.014,

    // Enriched locally
    expectedReturns: 0.13
  }

  const index = 1

  beforeEach(() => {
    $store = {
      state: {
        pool: {
          pool: {
            bonded_tokens: 1000
          }
        }
      }
    }

    wrapper = shallowMount(LiValidator, {
      localVue,
      propsData: {
        validator,
        index,
        showOnMobile: "returns"
      },
      mocks: {
        $store
      },
      stubs: [`router-link`]
    })
  })

  it(`has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should show the voting power`, () => {
    expect(wrapper.html()).toContain(`1.40%`)
  })

  it(`should gravatar when no avatar`, () => {
    wrapper.setProps({
      validator: {
        ...validator,
        avatarUrl: ""
      }
    })
    expect(wrapper.find("avatar-stub").exists())
  })
})
