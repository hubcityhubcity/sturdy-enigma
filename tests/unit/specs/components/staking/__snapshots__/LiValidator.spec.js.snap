// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiValidator has the expected html structure 1`] = `
<tr
  class="li-validator"
  data-moniker="mr_mounty"
>
  <td>
    2
  </td>
   
  <td
    class="hide-xs"
  >
    <div
      class="status-container"
    >
      <span
        class="validator-status active"
      >
        
        Active
      
      </span>
    </div>
  </td>
   
  <td
    class="data-table__row__info"
  >
    <img
      alt="validator logo for mr_mounty"
      class="li-validator-image"
      src="https://s3.amazonaws.com/img/path.jpg"
    />
     
    <div
      class="validator-info"
    >
      <h3
        class="li-validator-name"
      >
        
        mr_mounty
      
      </h3>
       
      <!---->
    </div>
  </td>
   
  <td
    class="hide-xs"
  >
    
    13.00%
  
  </td>
   
  <td
    class="hide-xs"
  >
    
    1.40%
  
  </td>
</tr>
`;
